### PR TITLE
Fix to maintain compatability with APS GE detectors

### PR DESCRIPTION
**Pull request by joelvbernier**
As ridiculous as it sounds, several years ago the staff at 1-ID tacitly decided to clobber the header on the GE files when they updated the firmware to 64-bit.  They kept a header with the same default length (8192 bytes), but stuffed it full of zeros; needless to say, this prevented fabio from being able to detect and read these modified GE files.  In the intervening time, _a lot_ of data has been measured with the GE detectors there (there is a solo Angio, and 4 GE41RT's arrayed as a composite WAXS detector), and they have not been amenable to fixing the header issue.  We've been working with my fork of fabio there since the change, which had fallen woefully out of date with your (canonical?) fork.  

So -- to maintain compatibility moving forward, and to simplify the build req's with my software package [HEXRD](https://github.com/HEXRD), I am sheepishly submitting this pull request for review.

Thanks!

### CODE REVIEW

**Comment by vallsv** [comment]
Hi,

Thanks a lot for the pull request.

It have to be reviewer seriously, but here is few comments:
- it would be good to provide a sample if possible (a small file that we could use for your test suite). We can take care of adding the test
- Else a pure generate test would be nice
- As we do not maintain the conda configuration, it would be better to remove the extra files related to conda. There is a dedicated project on conda-force
-  Finally i think there is a problem with line break in `GEimage`. It makes the review mostly impossible. Could it be possible to normalize it? Maybe WIndows/Linux issue?

Anyway thanks for sharing it with us. Empty header file is really interesting concept :-D That's a lot for inspiration.

File: fabio/GEimage.py [hunk 1]
```diff
@@ -1,332 +1,384 @@
...
+        """Read header into self.header and the data into self.data."""
+        if frame is None:
+            frame = 0
+        self.header = self.check_header()
+        self.resetvals()
+        infile = self._open(fname, "rb")
+        self.sequencefilename = fname
+        self._readheader(infile)
+
+        # obsolete for now # self.nframes = self.header['NumberOfFrames']
```
Comment (jonwright) [inline fabio/GEimage.py:273]: What if header was not corrupted? How about:
```
if self.header['NumberOfFrames'] == 0 : # struct.unpack("<H", b"\x00\x00")[0]
    file_size = os.stat(fname).st_size
    assert (numpy.remainder(file_size, self._BytesPerFrame)
                == self._HeaderNBytes), "GE file size is incorrect"
    self._nframes = file_size // self._BytesPerFrame
else:
    self._nframes = self.header['NumberOfFrames']

File: fabio/GEimage.py [hunk 2]
```diff
@@ -1,332 +1,384 @@
...
+        Returns
+        -------
+        None.
+
+        Notes
+        -----
+        The first image in the sequence 0; raises an exception if you give an
+        invalid image index, otherwise fills in self.data.
+        """
+        if(img_num >= self.nframes or img_num < 0):
```
[the remark this answers is not in the document]
Comment (jonwright) [reply fabio/GEimage.py:316]: Now I get it - the code was wrong before? How did it ever work ??

[the remark this answers is not in the document]
Comment (vallsv) [reply fabio/GEimage.py:316]: ^_^ was maybe designed to check if people use it.

[the remark this answers is not in the document]
Comment (joelvbernier) [reply fabio/GEimage.py:316]: Maybe I was the primary GEimage user?

File: fabio/GEimage.py (hunk 1 above)
Comment (joelvbernier) [reply fabio/GEimage.py:273]: That is a more elegant solution for sure -- I agree.

**Comment by joelvbernier** [comment]
@vallsv -- they initially got rid of the header wholesale, then put it back as all 0's so my old GUI could still load it...  all without informing anyone.  Ooof.  I have example images I can dig up -- they are, as you know, 8MB a frame.  Do we have space to add a test dataset with "old" (_i.e._, with header) and "new" (blanked header) images?  If so, should we do multiframe to test length?

**Comment by jonwright** [comment]
Mea culpa! I see I screwed up the '<=' business here :

https://github.com/silx-kit/fabio/commit/31562ed9035a5f5839336c84104f6e137c1fbe4f

Same error again today, ouch. Ken Evans did some debugging on this, but we were only using "next" to go through a series, and never found the other problem. I think he had it working to skip from the end of one file to the start of the next file with this patch :

https://github.com/silx-kit/fabio/commit/94f2221234d7dbebee68cbf0ebd061b3d33228b6

We had a testcase already, but it looks like a single frame only:

http://www.edna-site.org/pub/fabio/testimages/GE_...

For the "new header" you could copy that and overwrite the first bytes within a test setUp. If you use any of the file series things then it might be better to have a real multiframe test case to keep those working. For the case of all-zero headers it might be easy to generate it on the fly?

**Comment by kif** [comment]
We have no problem with the size of those frames. 
Provide us a few of each of those variants.

**Comment by joelvbernier** [comment]
[Sample_GE_Images.zip](https://github.com/silx-kit/fabio/files/4795499/Sample_GE_Images.zip)
Sample files attached, single frame CeO2 images.  Header status specifies in filename (intact is the original ADEPT header...).  LMK how it goes!

**Comment by vallsv** [comment]
Thanks a lot. I have uploaded the 2 images there http://www.silx.org/pub/fabio/testimages/

**Comment by vallsv** [comment]
It you don't care about cleaning the PR, or you don't know how to do it. Maybe we could work on the branch created by Jon. It have to be rebased to clean up things, and i can do it.

**Comment by jonwright** [comment]
@vallsv I think you are asking @joelvbernier? I will be very happy if you can figure out how to clean up my mess. I will try to rename this silx-kit remote as upstream to avoid future mishaps.

**Comment by vallsv** [comment]
I was asking both of you in fact. Anyway what you do is not a mess, it's just a branch. It can be removed with the web interface without problem. But we also could use it to rework the content of this PR, that's why i asked.

**Comment by vallsv** [comment]
I close this PR as it is supperceded by #396

**Pull request closed without merging**